### PR TITLE
Stubbed UI to create line items

### PR DESCRIPTION
**Pull request by deammer**
### Changes

- allow users to create and edit line items in the UI
- added a `lineItem` query resolver

### Video

https://user-images.githubusercontent.com/3411183/113091576-16032880-91a1-11eb-9c34-cc7442532197.mov

### CODE REVIEW

File: frontend/src/data/constants.ts [hunk 1]
```diff
@@ -60,6 +61,13 @@ export const PALLET_PAYMENT_STATUS_OPTIONS = [
   },
 ]
 
+export const LINE_ITEM_CONTAINER_OPTIONS = [
+  { label: 'Not set', value: LineItemContainerType.Unset },
+  { label: 'Box', value: LineItemContainerType.Box },
+  { label: 'Bulk bag', value: LineItemContainerType.BulkBag },
+  { label: 'Full pallet', value: LineItemContainerType.FullPallet },
+] as const
```
Comment (deammer) [inline frontend/src/data/constants.ts:69]: I recently learned that `as const` forces a variable to be read-only! This means that the array can't be modified by methods like `push()`.

File: frontend/src/pages/offers/PalletsEditor.tsx [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+          <div className="flex h-full w-full items-center justify-center bg-gray-50 text-gray-500">
+            <p>← Select a line item</p>
+          </div>
+        )}
+        {selectedPalletId != null && selectedLineItemId != null && (
+          <div className="h-full w-full p-4">
+            <LineItemForm lineItemId={selectedLineItemId} />
+          </div>
+        )}
+      </div>
```
Comment (deammer) [inline frontend/src/pages/offers/PalletsEditor.tsx:126]: WIP code. Will break it up into components once the features have been finalized.

File: frontend/src/pages/offers/ViewOfferPage.tsx [hunk 3]
```diff
@@ -132,20 +132,27 @@ const ViewOfferPage: FunctionComponent = () => {
...
-                      {shipment.shipment.sendingHub.name} →{' '}
-                      {shipment.shipment.receivingHub.name}
-                    </p>
-                  </>
-                )}
-              </ReadOnlyField>
-              <ReadOnlyField label="Group">
-                {sendingGroup?.group.name}
-              </ReadOnlyField>
+              <div className="flex space-x-8">
```
Comment (deammer) [inline frontend/src/pages/offers/ViewOfferPage.tsx:135]: Screen Shot 2021-03-30 at 9 42 07 PM

File: frontend/src/data/constants.ts (hunk 1 above)
Comment (bion) [reply frontend/src/data/constants.ts:69]: 🤯

File: frontend/src/pages/offers/PalletsEditor.tsx [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+  pallets: OfferQuery['offer']['pallets']
+  initiateDeletePallet: (palletId: number) => void
+}
+
+/**
+ * This is a 2-column UI with a list of pallets on the left and details aobut
+ * each line-item on the right.
+ */
+const PalletsEditor: FunctionComponent<Props> = ({ pallets }) => {
+  // TODO move this to the URL?
```
Comment (bion) [inline frontend/src/pages/offers/PalletsEditor.tsx:20]: thought (non-blocking): yep, that would be good so people can share a link to a specific pallet

File: frontend/src/pages/offers/PalletsEditor.tsx [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+        <ul className="divide-y divide-gray-100">
+          {pallets.map((pallet) => (
+            <li
+              key={pallet.id}
+              className="bg-white"
+              onClick={() => selectPallet(pallet.id)}
+            >
+              <div
+                className={cx('p-4 bg-white border-l-4 border-transparent', {
+                  ' border-blue-500': selectedPalletId === pallet.id,
```
Comment (bion) [inline frontend/src/pages/offers/PalletsEditor.tsx]: nit: 
```suggestion
                  'border-blue-500': selectedPalletId === pallet.id,
```

File: src/resolvers/line_items.ts [hunk 6]
```diff
@@ -13,12 +13,23 @@ import {
...
 import validateEnumMembership from './validateEnumMembership'
 import validateUris from './validateUris'
 
+const lineItem: QueryResolvers['lineItem'] = async (_, { id }, context) => {
+  const lineItem = await getLineItemWithParentAssociations(id)
+
+  if (!lineItem) {
+    throw new UserInputError(`LineItem ${id} does not exist`)
+  }
+
```
Comment (bion) [inline src/resolvers/line_items.ts:32]: issue (blocking): this query needs to be authorized. Either authorize it using [`authorizeOfferQuery`](https://github.com/distributeaid/shipment-tracker/blob/saga/src/resolvers/offer_authorization.ts#L83) (you're already using `getLineItemWithParentAssociations` so the offer is available) or file an issue for authorizing it and I can pick it up.

File: src/resolvers/line_items.ts [hunk 7]
```diff
@@ -13,12 +13,23 @@ import {
...
   LineItemUpdateInput,
   MutationResolvers,
+  QueryResolvers,
 } from '../server-internal-types'
 import getPalletWithParentAssociations from './getPalletWithParentAssociations'
 import { authorizeOfferMutation } from './offer_authorization'
 import validateEnumMembership from './validateEnumMembership'
 import validateUris from './validateUris'
 
+const lineItem: QueryResolvers['lineItem'] = async (_, { id }, context) => {
```
Comment (bion) [inline src/resolvers/line_items.ts:26]: suggestion (blocking): please add a test case for at least the successful case for this endpoint

**Review by bion — APPROVED** [review]

File: frontend/src/pages/offers/PalletsEditor.tsx (hunk 4 above)
Comment (deammer) [reply frontend/src/pages/offers/PalletsEditor.tsx:20]: Yeah exactly. In a future PR!

**Pull request merged**